controllers/fluentbit_controller.go

```diff
@@ -78,206 +78,59 @@ func (r *FluentBitReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-		} else {
-			return "/var/lib/docker/containers"
-		}
-	}()
+func (r *FluentBitReconciler) mutate(ds *appsv1.DaemonSet, fb logging.FluentBit) controllerutil.MutateFn {
+	logPath := r.getContainerLogPath(fb)
+	expected := operator.MakeDaemonSet(fb, logPath)
 
-	ds := appsv1.DaemonSet{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      fb.Name,
-			Namespace: fb.Namespace,
-			Labels:    fb.Labels,
-		},
-		Spec: appsv1.DaemonSetSpec{
-			Selector: &metav1.LabelSelector{
-				MatchLabels: fb.Labels,
-			},
-			Template: corev1.PodTemplateSpec{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      fb.Name,
-					Namespace: fb.Namespace,
-					Labels:    fb.Labels,
-				},
-				Spec: corev1.PodSpec{
-					ServiceAccountName: fb.Name,
-					Volumes: []corev1.Volume{
-						{
-							Name: "varlibcontainers",
-							VolumeSource: corev1.VolumeSource{
-								HostPath: &corev1.HostPathVolumeSource{
-									Path: logPath,
-								},
-							},
-						},
-						{
-							Name: "config",
-							VolumeSource: corev1.VolumeSource{
-								Secret: &corev1.SecretVolumeSource{
-									SecretName: fb.Spec.FluentBitConfigName,
-								},
-							},
-						},
-						{
-							Name: "varlogs",
-							VolumeSource: corev1.VolumeSource{
-								HostPath: &corev1.HostPathVolumeSource{
-									Path: "/var/log",
-								},
-							},
-						},
-						{
-							Name:         "positions",
-							VolumeSource: fb.Spec.PositionDB,
-						},
-					},
-					Containers: []corev1.Container{
-						{
-							Name:            "fluent-bit",
-							Image:           fb.Spec.Image,
-							ImagePullPolicy: fb.Spec.ImagePullPolicy,
-							Ports: []corev1.ContainerPort{
-								{
-									Name:          "metrics",
-									ContainerPort: 2020,
-								},
-							},
-							VolumeMounts: []corev1.VolumeMount{
-								{
-									Name:      "varlibcontainers",
-									ReadOnly:  true,
-									MountPath: logPath,
-								},
-								{
-									Name:      "config",
-									ReadOnly:  true,
-									MountPath: "/fluent-bit/config",
-								},
-								{
-									Name:      "positions",
-									MountPath: "/fluent-bit/tail",
-								},
-								{
-									Name:      "varlogs",
-									ReadOnly:  true,
-									MountPath: "/var/log/",
-								},
-							},
-						},
-					},
-					Tolerations: fb.Spec.Tolerations,
-				},
-			},
-		},
-	}
-
-	// Mount Secrets
-	for _, secret := range fb.Spec.Secrets {
-		ds.Spec.Template.Spec.Volumes = append(ds.Spec.Template.Spec.Volumes, corev1.Volume{
-			Name: secret,
-			VolumeSource: corev1.VolumeSource{
-				Secret: &corev1.SecretVolumeSource{
-					SecretName: secret,
-				},
-			},
-		})
-		ds.Spec.Template.Spec.Containers[0].VolumeMounts = append(ds.Spec.Template.Spec.Containers[0].VolumeMounts, corev1.VolumeMount{
-			Name:      secret,
-			ReadOnly:  true,
-			MountPath: fmt.Sprintf("/fluent-bit/secrets/%s", secret),
-		})
+	return func() error {
+		ds.Labels = expected.Labels
+		ds.Spec = expected.Spec
```

Comment:
Simply assign the entire expected spec to ds.spec is kind of dangerous. I'll merge this PR for now because you mentioned no infinite reconcile loop happens. It's better to make sure of this by adding debug logs.
You can refine this part later maybe. Junot's implementation is a good reference: https://github.com/kubesphere/kube-events/blob/master/pkg/controllers/kubeeventsruler_controller.go#L336